drs_downloader/clients/terra.py

```diff
@@ -0,0 +1,108 @@
+import subprocess
+import tempfile
+from pathlib import Path
+
+import aiofiles
+import aiohttp
+import logging
+
+from drs_downloader import MB
+from drs_downloader.models import DrsClient, DrsObject, AccessMethod, Checksum
+
+logger = logging.getLogger(__name__)
+
+
+class TerraDrsClient(DrsClient):
+    """
+    Calls the terra DRS server.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.endpoint = "https://us-central1-broad-dsde-prod.cloudfunctions.net/martha_v3"
+        self.token = self._get_auth_token()
+
+    @staticmethod
+    def _get_auth_token() -> str:
+        """Get Google Cloud authentication token.
+        User must run 'gcloud auth login' from the shell before starting this script.
+
+        Returns:
+            str: auth token
+        """
+        token_command = "gcloud auth print-access-token"
+        cmd = token_command.split(' ')
+        token = subprocess.check_output(cmd).decode("ascii")[0:-1]
+        assert token, "No token retrieved."
+        return token
+
+    async def download_part(self, drs_object: DrsObject, start: int, size: int, destination_path: Path) -> Path:
+        try:
+            headers = {'Range': f'bytes={start}-{size}'}
+
+            file_name = destination_path / f'{drs_object.name}.{start}.{size}.part'
+            async with aiohttp.ClientSession(headers=headers) as session:
+                async with session.get(drs_object.access_methods[0].access_url) as request:
+                    file = await aiofiles.open(file_name, 'wb')
+                    self.statistics.set_max_files_open()
+                    async for data in request.content.iter_any():  # uses less memory
+                        await file.write(data)
```

Comment:
Does L 48-49 write each packet of data as it is read from the google bucket, or is there an opportunity to make more optimizations here?